zebra-chain/src/value_balance.rs

```diff
@@ -30,8 +30,13 @@ where
     pub fn remaining_transaction_value(&self) -> Result<Amount<NonNegative>, Error> {
         // This rule checks the transparent value balance minus the sum of the sprout,
         // sapling, and orchard value balances in a transaction is nonnegative.
-        (self.transparent - (self.sprout + self.sapling + self.orchard)?)?
-            .constrain::<NonNegative>()
+        // Calculated in Zebra by negating the sum of the transparent, sprout,
+        // sapling, and orchard value balances as specified in
+        // https://zebra.zfnd.org/dev/rfcs/0012-value-pools.html#definitions
+        let amount = -(self.transparent + self.sprout + self.sapling + self.orchard)?
+            .constrain::<NegativeAllowed>()?;
+
+        amount.constrain::<NonNegative>()
```

Comment:
Why does this code constrain twice?

It might help to think about what's happening in each line.

Reply:
The problem is that we had a `Neg` implementation only for `Amount<NegativeAllowed>`. In here the sum can be negative but the negation of it can't, that was the reason of the two constrain calls.

I replaced the `Neg` implementation with a generic one that returns `Result` at https://github.com/ZcashFoundation/zebra/pull/2569/commits/a012b582ba64ef5085df3a9ac9a90da56a994d2a and applied it.